wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheMonitorImpl.java

```diff
@@ -0,0 +1,919 @@
+import io.lettuce.core.RedisCommandExecutionException;
+import io.lettuce.core.RedisConnectionException;
+import io.lettuce.core.RedisException;
+import io.lettuce.core.RedisURI;
+import io.lettuce.core.api.StatefulRedisConnection;
+import io.lettuce.core.codec.ByteArrayCodec;
+import io.lettuce.core.resource.ClientResources;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.AwsWrapperProperty;
+import software.amazon.jdbc.util.telemetry.TelemetryCounter;
+import software.amazon.jdbc.util.telemetry.TelemetryFactory;
+import software.amazon.jdbc.util.telemetry.TelemetryGauge;
+
+/**
+ * This class uses a background thread to monitor cache cluster health and manage state transitions.
+ *
+ * Implements a three-state machine (HEALTHY → SUSPECT → DEGRADED) with proactive health checks
+ * that only run when clusters are in SUSPECT or DEGRADED states.
+ */
+public class CacheMonitorImpl implements CacheMonitor {
+
+  private static final Logger LOGGER = Logger.getLogger(CacheMonitorImpl.class.getName());
+
+  // Singleton instance
+  private static volatile CacheMonitorImpl instance;
+  private static final Object INSTANCE_LOCK = new Object();
+
+  private static final long THREAD_SLEEP_WHEN_INACTIVE_MILLIS = 100;
+  private static final int CACHE_HEALTH_CHECK_INTERVAL = 5;
+  private static final int CACHE_CONSECUTIVE_SUCCESS_THRESHOLD = 3;
+  private static final int CACHE_CONSECUTIVE_FAILURE_THRESHOLD = 3;
+
+  // Dedicated ping connections - one for RW, one for RO if different
+  private static volatile StatefulRedisConnection<byte[], byte[]> pingConnectionRw;
+  private static volatile StatefulRedisConnection<byte[], byte[]> pingConnectionRo;
+  private static final ReentrantLock PING_RW_LOCK = new ReentrantLock();
+  private static final ReentrantLock PING_RO_LOCK = new ReentrantLock();
+
+  // Track RW and RO endpoints for ping connection management
+  private static volatile String cachedRwEndpoint;
+  private static volatile String cachedRoEndpoint;
+
+  // Track if monitor thread has been started
+  private static volatile boolean monitorThreadStarted = false;
+  private static final Object THREAD_START_LOCK = new Object();
+
+  // Configuration properties
+  public static final AwsWrapperProperty CACHE_IN_FLIGHT_WRITE_SIZE_LIMIT =
+      new AwsWrapperProperty(
+          "cacheInFlightWriteSizeLimitBytes",
+          String.valueOf(50 * 1024 * 1024), // 50 MB
+          "Maximum in-flight write size in bytes before triggering degraded mode.");
+
+  public static final AwsWrapperProperty CACHE_HEALTH_CHECK_IN_HEALTHY_STATE =
+      new AwsWrapperProperty(
+          "cacheHealthCheckInHealthyState",
+          "false",
+          "Whether to run health checks (pings) in healthy state.");
+
+  private final Map<String, EndpointHealthState> endpointStates = new ConcurrentHashMap<>();
+  private final Map<String, ClusterHealthState> clusterStates = new ConcurrentHashMap<>();
```

Comment:
As we discussed, remove endpoint specific map. Keep a single source of truth (cluster level map, with the rw/ro endpoint name as the key, and read/write endpoint information as the value). 

This will make the logic a lot cleaner and easier to read.